workshop/session_3/09_biopython_sequences.ipynb

```diff
@@ -0,0 +1,856 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 141,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Alignment(seqA='ACGT-', seqB='ACG-C', score=3.0, start=0, end=5)\n",
+      "Alignment(seqA='ACGT', seqB='ACGC', score=3.0, start=0, end=4)\n"
+     ]
+    }
+   ],
+   "source": [
+    "from Bio import pairwise2\n",
+    "\n",
+    "# Notice we can supply either Seq's or strings directly.\n",
+    "global_alignments = pairwise2.align.globalxx(Seq(\"ACGT\"), \"ACGC\")\n",
+    "for ga in global_alignments:\n",
+    "    print(ga)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": false
+   },
+   "source": [
+    "This representation is not really user friendly. \n",
+    "\n",
+    "Let's try to make it look better."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 142,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "ACGT-\n",
+      "|||  \n",
+      "ACG-C\n",
+      "  Score=3\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Print the first available alignment. \n",
+    "# The * converts the Alignment objects into a list of parameters required by the format_alignment function.\n",
+    "print(pairwise2.format_alignment(*global_alignments[0]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": false
+   },
+   "source": [
+    "Similar to the global alignment, we can use `pairwise2` to perform local alignment of 2 sequences."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 143,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "1 ACG\n",
+      "  |||\n",
+      "1 ACG\n",
+      "  Score=3\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "local_alignments = pairwise2.align.localxx(Seq(\"ACGT\"), \"ACGC\")\n",
+    "print(pairwise2.format_alignment(*local_alignments[0]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "editable": false
+   },
+   "source": [
+    "Multiple sequence alignment is suitable for aligning 3 or more sequences at a time. \n",
+    "\n",
+    "*Note: for MSA, sequences are expected to be of the same length.* "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 144,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from Bio.Align import MultipleSeqAlignment, AlignInfo\n",
+    "from Bio.SeqRecord import SeqRecord\n",
+    "\n",
+    "seqs = [\n",
+    "    \"ACGTACGT\",\n",
+    "    \"ACGTGCGC\",\n",
+    "    \"ACGTA--T\",\n",
+    "    \"CCGTACGG\",\n",
+    "    \"A-GTACCC\",\n",
+    "    \"ACGTA--T\",\n",
+    "    \"CTG-ACG-\",\n",
+    "    \"AGGTACG-\"\n",
+    "]\n",
+    "aligned = MultipleSeqAlignment([SeqRecord(Seq(s)) for s in seqs])"
```

Comment:
The `SeqRecord` class is used below but it isn't really introduced. We should add a brief explanation of the difference between `Seq` and `SeqRecord`

Reply:
We never introduce the list comprehension syntax. I would suggest creating the list separately in this case rather than try to introduce a new concept

Reply:
You're absolutely right, moved SeqRecord from FASTA description to this section. 
Got rid of the list comprehension.